arch:arm:core:aarch32: Basic ARMv7-A / Cortex-A support

ibirnbaum opened the pull request:
This package contains work-in-progress versions of all the components required to get initial support of ARMv7-A / 32-bit Cortex-A CPUs up and running. For now, it's intended as a basis for discussion, in particular regarding the MMU implementation and the resulting changes to the linker command files, as well as the proper integration of each of the new or modified components (e.g. proper naming / use / location of Kconfig items).

The components included are:
- Basic context switching and interrupt handling. This is basically completely parallel to the Cortex-R implementation, i.e. every #ifdef CONFIG_ARMV7_R has been OR'ed with CONFIG_ARMV7_A, and every CONFIG_CPU_CORTEX_R has been OR'ed with CONFIG_CPU_AARCH32_CORTEX_A. Interrupt handling uses the existing GICv1 driver. System timing is provided by the ARM architected timer.
- However, the timer driver required the addition of a workaround for an erratum which affects the Cortex-A9 CPU. A description of the erratum and a reference to the corresponding documentation by ARM is provided in the respective commit.
- Basic support of the ARMv7 MMU. Supports static memory area configuration at boot time as well as dynamic configuration based on run-time calls to arch_mem_map. No userspace support yet, no stack guard pages yet, all other minor caveats are described in the respective commit.
- SoC/DT and other related data for the Xilinx Zynq-7000 SoC series. This SoC family should be a suitable candidate for a first aarch32 Cortex-A target, as drivers for all relevant peripherals are available (UART, Ethernet, with slight modifications of the existing ZynqMP implementations).
- Support for the Cortex-A's On-Chip Memory (OCM) so that the location of this block of memory can be configured and data can be explicitly placed in the corresponding section. This memory can be used, for example, for uncached device DMA.
- The "qemu_cortex_a9" board definition, which simulates the Xilinx Zynq-7000. The QEMU simulation includes all features relevant for testing & QA. NOTE: at least the QEMU version that currently comes with Ubuntu 20.04, qemu-system-arm version 4.2.1, crashes with a Floating Point Exception when starting the simulation, although the FPU isn't used within the target. I've manually installed a different version of QEMU that worked before, I'll try to provide more info which QEMU versions work properly. I believe this problem to be a QEMU bug.
- A fix for interrupt-driven operation of the Xilinx PS UART driver.
- An update to the GEM Ethernet driver so that it supports the Zynq-7000 SoCs in addition to the ZynqMP SoCs.

What is NOT included (that hasn't been mentioned already above):
- FPU support. I've removed my old FPU support implementation as recent discussions hinted at an implementation for the Cortex-R VFPv3-D16 FPU being in the works, this should be expandable to the Cortex-A's FVPv3-D32 FPU.

ibirnbaum committed:
arch: arm: core: aarch32: Updated brief file description

Updated brief file description so that it also mentions the aarch32
Cortex-A CPUs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
arch: arm: core: aarch32: Add ARMv7-A/Cortex-A(9) related Kconfig items

Add the ARMV7_A, CPU_AARCH32_CORTEX_A and CPU_CORTEX_A9 configuration
items.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
arch: arm: core: aarch32: Updated brief file description

Updated brief file description so that it also mentions the aarch32
Cortex-A CPUs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
arch: arm: core: aarch32: limit ACTLR register access to Cortex-R

The configuration bits ATCMPCEN, B0TCMPCEN and B1TCMPCEN in the ACTLR
register referenced in the function z_arm_tcm_disable_ecc are only de-
fined for Cortex-R CPUs. For Cortex-A CPUs, those bits are declared
as reserved.

Comp.: https://arm-software.github.io/CMSIS_5/Core_A/html/group__CMSIS__ACTLR.html

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
arch: arm: core: aarch32: introduce basic ARMv7 MMU support

An initial implementation for memory management using the ARMv7 MMU.
A single L1 translation table for the whole 4 GB address space is al-
ways present, a configurable number of L2 page tables are linked to
the L1 table based on the static memory area configuration at boot
time, or whenever arch_mem_map/arch_mem_unmap are called at run-time.

Currently, a CPU with the Multiprocessor Extensions and execution at
PL1 are always assumed. Userspace-related features or thread stack
guard pages are not yet supported. Neither are LPAE, PXN or TEX re-
mapping. All mappings are currently assigned to the same domain. Re-
garding the permissions model, access permissions are specified using
the AP[2:1] model rather than the older AP[2:0] model, which, accor-
ding to ARM's documentation, is deprecated and should no longer be
used. The newer model adds some complexity when it comes to mapping
pages as unaccessible (the AP[2:1] model doesn't support explicit
specification of "no R, no W" permissions, it's always at least "RO"),
this is accomplished by invalidating the ID bits of the respective
page's PTE.

Includes sources, Kconfig integration, adjusted CMakeLists and the
modified linker command file (proper section alignment!).

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
arch: arm: core: aarch32: enable ARMv7-R/Cortex-R code for ARMv7-A/Cortex-A

Modify #ifdefs so that any code that is compiled if CONFIG_ARMV7_R is
set is also compiled if CONFIG_ARMV7_A is set.
Modify #ifdefs so that any code that is compiled if CONFIG_CPU_CORTEX_R
is set is also compiled if CONFIG_CPU_AARCH32_CORTEX_A is set.
Modify source dir inclusion in CMakeLists.txt accordingly.

Brief file descriptions have been updated to include Cortex-A whereever
only Cortex-M and Cortex-R were mentioned so far.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
cmake: add the gcc -mcpu parameter value for the Cortex-A9 CPU

Add the -mcpu=cortex-a9 option as part of the basic aarch32 Cortex-A
CPU support.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: timer: arm_arch_timer: Workaround for Cortex-A9 erratum 740657

Modification of the ARM architected timer driver and its configuration
data in order to address an erratum which exists at least in the Cor-
tex-A9 CPU, and which can also be observed in the QEMU implementation
of the Cortex-A9.

Comp.: ARM Cortex-A9 processors Software Developer Errata Notice
ARM document ID032315
Erratum 740657

This erratum causes a spurious interrupt pending indication with the
interrupt controller if no new compare value is written within the
timer ISR before the interrupt is cleared. This is usually the case
in tickless mode. If the spurious interrupt is not prevented, the
timer ISR will be called twice, but on second execution, the pending
flag is not set within the timer's register space. Not handling this
issue will lead to erratic tick announcements to the kernel.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
soc: arm: dts: arm: xilinx: Zynq-7000 SoC init code, device tree data

Add SoC-specific code, the basic device tree and Kconfig data as well
as the corresponding linker command file for the Xilinx Zynq-7000
family of SoCs. This SoC - either as a QEMU simulation or on actual
hardware such as the Avnet/Digilent ZedBoard - is suitable as an ini-
tial target for the ARMv7 Cortex-A support.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: Zynq-7000 support for the Xilinx GEM driver

Add support for the Xilinx Zynq-7000 SoC family to this driver. This
includes some SoC-specific register accesses when setting an updated
TX clock divider, also, the device tree binding now supports higher
MDC clock divisor values when the current target SoC is a Zynq rather
than a ZynqMP.

With regards to the use of this driver in a QEMU simulation of the
Zynq-7000, the Kconfig file is modified so that the driver is not
enabled unless QEMU networking is set to Ethernet mode.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
arch: arm: core: aarch32: support for On-Chip Memory (OCM) areas in aarch32

Introduce support for On-Chip Memory (OCM) areas in aarch32 Cortex-A
SoCs. This memory area is suitable for, amongst other things, un-
cached device DMA. The OCM memory can reside either at address zero
or at a high memory address, with configuration of the memory's lo-
cation similar to the configuration of the location of the exception
vectors. In both cases, it must be assured that the low/high OCM
area declaration doesn't overlap with the exception vectors memory
page at either the low/high address. Therefore, all possible OCM
location options shall be described in the SoC's device tree, the
current location shall be indicated via a "chosen" entry in the
target board's device tree.

This commit contains the necessary changes to the linker-related
files as well as the static region declaration for the MMU in-
itialization in the Zynq-7000 SoC-specific init code.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
boards: arm: QEMU target for Cortex-A9 simulation

The qemu_cortex_a9 target simulates the Xilinx Zynq-7000
in a single core configuration with 512 MB of RAM. Supported
peripherals are the Zynq's two UARTs, of which one is activated
for console use, and the two Ethernet controllers, of which
GEM0 can be linked to the local zeth interface.

This target uses Xilinx' fork of QEMU, which is already being
used by the qemu_cortex_r5 target. Although being labelled
'arm64', this version of QEMU is also capable of emulating the
Zynq-7000 based on a binary device tree file. While regular
QEMU provides a Zynq-based machine, this implementation is
buggy and/or incomplete.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: serial: xlnx_xuartps: enable this driver for Zynq-7000 series SoCs

Enable this driver to be used in conjunction with Xilinx Zynq-7000
series SoCs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: serial: xlnx_xuartps: Fix interrupt-driven operation

Fixes interrupt-driven operation. With the previous way of
handling the TX FIFO and the interrupt flags, the operation
of the UART was prone to stalling when using it as the
console I/O device.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

carlocaione commented:
On a first look LGTM (great work BTW). Can you fix the buildkite errors though?

bbolen commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+		reg_val |= ARM_MMU_TTBR_IRGN1_BIT_MP_EXT_ONLY;
+	}
+
+	__asm__ __volatile__("mcr p15, 0, %0, c2, c0, 0" : : "r"(reg_val));
```
I think you can use __get_TTBR0() from the cmsis

ibirnbaum replied:
CMSIS integration is my goal, yet, with CMSIS being an external component, I haven't yet figured out where to commit the CPU-specific header core_ca9.h, similar to the already existing core_cr5.h, as CMSIS header inclusion entry point.

stephanosio replied:
CMSIS-Core(A) has already been integrated.

See https://github.com/zephyrproject-rtos/cmsis/tree/master/CMSIS/Core_A/Include

ibirnbaum replied:
When looking for the CMSIS, I only looked within zephyrproject-rtos/zephyr and in ARM-software/CMSIS_5, but not zephyrproject-rtos/... D'oh!

ibirnbaum replied:
the MMU code has been modified to use CMSIS whereever supported, the only exceptions being TTBR1 and TTBCR, which aren't covered by cmsis_cp15.h.

bbolen commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+
+	/* Write DACR -> all domains to client = 01b. */
+	reg_val = ARM_MMU_DACR_ALL_DOMAINS_CLIENT;
+	__asm__ __volatile__("mcr p15, 0, %0, c3, c0, 0" : : "r"(reg_val));
```
__set_DACR() from cmsis?

ibirnbaum replied:
done

bbolen commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+	invalidate_tlb_all();
+
+	/* Enable the MMU and Cache in SCTLR */
+	__asm__ __volatile__("mrc p15, 0, %0, c1, c0, 0" : "=r"(reg_val));
```
__get_SCTLR()

ibirnbaum replied:
done

bbolen commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+	reg_val |= ARM_MMU_SCTLR_ICACHE_ENABLE_BIT;
+	reg_val |= ARM_MMU_SCTLR_DCACHE_ENABLE_BIT;
+	reg_val |= ARM_MMU_SCTLR_MMU_ENABLE_BIT;
+	__asm__ __volatile__("mcr p15, 0, %0, c1, c0, 0" : : "r"(reg_val));
```
__set_SCTLR

ibirnbaum replied:
done

bbolen commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+{
+	uint32_t reg_val = 0; /* 0 = opc2 = invalidate entire TLB */
+	__asm__ __volatile__("mcr p15, 0, %0, c8, c7, 0" : : "r"(reg_val));
+	__asm__ __volatile__("dsb sy");
```
__DSB()

ibirnbaum replied:
done

bbolen commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+	uint32_t reg_val = 0; /* 0 = opc2 = invalidate entire TLB */
+	__asm__ __volatile__("mcr p15, 0, %0, c8, c7, 0" : : "r"(reg_val));
+	__asm__ __volatile__("dsb sy");
+	__asm__ __volatile__("isb");
```
__ISB()

ibirnbaum replied:
done

bbolen commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+static void invalidate_tlb_all(void)
+{
+	uint32_t reg_val = 0; /* 0 = opc2 = invalidate entire TLB */
+	__asm__ __volatile__("mcr p15, 0, %0, c8, c7, 0" : : "r"(reg_val));
```
__set_TLBIALL(0)

ibirnbaum replied:
done

ibirnbaum commented on arch/arm/core/aarch32/mmu/arm_mmu.c:
```diff
@@ -0,0 +1,800 @@
+	l2_page_tables_status[arm_mmu_l2_next_free_table].entries = 0;
+	l2_page_table = &l2_page_tables[arm_mmu_l2_next_free_table];
+
+	if (--arm_mmu_l2_tables_free > 0)
```
Correct, that wasn't a clever shortcut...

ibirnbaum commented on soc/arm/xilinx_zynq7000/soc.c:
```diff
@@ -0,0 +1,83 @@
+ */
+static int soc_xlnx_zynq7000_init(const struct device *arg)
+{
+	uint32_t reg_val;
```
I'm leaving this in for now - I already had an implementation in here that writes the registers which configure where the 256 kB of OCM are located, based on the respective 'chosen' entry in the device tree. The power-on configuration is 3/4 of the OCM starting at address 0, 1/4 at 0xFFFC0000. The attempt to move the 64k located at the high address to the low address currently crashes the system, as we neither have proper SMP yet nor have I looked into how to shut down the 2nd core in non-SMP mode yet. The 'wait for event' loop code the 2nd core idles in after power on is located in that particular 64k block that's located at the high address by default. Therefore, I removed this code for now, but it shall eventually return one day.

ibirnbaum commented on drivers/serial/uart_xlnx_ps.c:
```diff
@@ -845,20 +842,19 @@ static int uart_xlnx_ps_fifo_fill(const struct device *dev,
-		reg_val = sys_read32(reg_base + XUARTPS_SR_OFFSET);
+
+	sys_write32(XUARTPS_IXR_TXEMPTY, reg_base + XUARTPS_IDR_OFFSET);
+	while (size) {
```
done

ibirnbaum commented on drivers/serial/uart_xlnx_ps.c:
```diff
@@ -845,20 +842,19 @@ static int uart_xlnx_ps_fifo_fill(const struct device *dev,
+	sys_write32(XUARTPS_IXR_TXEMPTY, reg_base + XUARTPS_IDR_OFFSET);
+	while (size) {
+		while ((sys_read32(reg_base + XUARTPS_SR_OFFSET) & XUARTPS_SR_TXFULL) != 0) {}
+		sys_write32((uint32_t)(tx_data[data_iter++] & 0xFF), reg_base + XUARTPS_FIFO_OFFSET);
```
done

ibirnbaum commented on soc/arm/xilinx_zynq7000/soc.c:
```diff
@@ -0,0 +1,83 @@
+			      MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),
+	/* ARM Arch timer, GIC are covered by the MPCore mapping */
+/* UARTs */
+#if DT_NODE_HAS_STATUS(DT_INST(0, xlnx_xuartps), okay)
```
done

ibirnbaum commented on boards/arm/qemu_cortex_a9/Kconfig.defconfig:
```diff
@@ -0,0 +1,72 @@
+	
+if SERIAL
+
+	config SERIAL_HAS_DRIVER
```
removed that item

ibirnbaum commented on boards/arm/qemu_cortex_a9/Kconfig.defconfig:
```diff
@@ -0,0 +1,72 @@
+	config SERIAL_HAS_DRIVER
+	default y
+
+	config UART_INTERRUPT_DRIVEN
```
That's where that item is now. Yet I'm currently wondering how I came up with the combination of CONSOLE=y, INTERRUPT_DRIVEN=n... More testing of the QEMU target is coming up anyways.

ibirnbaum replied:
The config files of the QEMU A9 target have been re-worked.

ibirnbaum commented on boards/arm/qemu_cortex_a9/qemu_cortex_a9.dts:
```diff
@@ -0,0 +1,56 @@
+		reg = <0x00000000 DT_SIZE_M(256)>;
+	};
+
+	aliases {
```
... aaaaand it's gone!

ibirnbaum commented on boards/arm/qemu_cortex_a9/qemu_cortex_a9_defconfig:
```diff
@@ -0,0 +1,24 @@
+CONFIG_CONSOLE=y
+CONFIG_NETWORKING=y
+CONFIG_NET_QEMU_ETHERNET=y
+CONFIG_GPIO=n
```
done

ibirnbaum commented on include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -84,10 +85,11 @@ _region_min_align = 4;
+    DT_REGION_FROM_NODE_STATUS_OKAY(OCM, rw, DT_CHOSEN(zephyr_ocm))
     /* Used by and documented in include/linker/intlist.ld */
-    IDT_LIST (wx) : ORIGIN = (RAM_ADDR + RAM_SIZE), LENGTH = 2K
+    IDT_LIST  (wx) : ORIGIN = (RAM_ADDR + RAM_SIZE), LENGTH = 2K
```
not sure what happened there, but it's fixed...

ibirnbaum commented on include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -84,10 +85,11 @@ _region_min_align = 4;
 {
-    FLASH    (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
-    SRAM     (wx) : ORIGIN = RAM_ADDR, LENGTH = RAM_SIZE
+    FLASH     (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
```
dito

ibirnbaum commented on drivers/timer/Kconfig:
```diff
@@ -121,6 +121,18 @@ DT_COMPAT_ARM_V7M_SYSTICK := arm,armv7m-systick
 
+config ARM_ARCH_TIMER_ERRATUM_740657
+	bool "ARM architected timer is affected by ARM erratum 740657"
+	default n
```
done

ibirnbaum commented on include/linker/sections.h:
```diff
@@ -51,6 +51,10 @@
+
+#define _OCM_DATA_SECTION_NAME	.ocm_data
+#define _OCM_BSS_SECTION_NAME		.ocm_bss
+#define _OCM_NOINIT_SECTION_NAME	.ocm_noinit
```
The Ethernet driver places its DMA area in the bss section, noinit is not really required. The availability of those three sections is simply based on what other On-Chip memory types that are already supported offer.

ibirnbaum replied:
Removed the noinit-related define, as this actually wasn't considered in the linker command file.

ibirnbaum commented on drivers/timer/arm_arch_timer.c:
```diff
@@ -39,7 +57,28 @@ static void arm_arch_timer_compare_isr(const void *arg)
+		 * is being processed before the interrupt is acknowledged
+		 * by the handling interrupt controller.
+		 */
+		arm_arch_timer_set_compare(0xFFFFFFFFFFFFFFFFLLU);
```
done

ibirnbaum commented on include/arch/arm/aarch32/cortex_a_r/timer.h:
```diff
@@ -63,6 +63,30 @@ static ALWAYS_INLINE void arm_arch_timer_set_compare(uint64_t val)
 	sys_write32(ctrl, TIMER_REG_GET(TIMER_CTRL));
 }
 
+#if defined(CONFIG_ARM_ARCH_TIMER_ERRATUM_740657) || defined(CONFIG_QEMU_TARGET)
```
I did actually observe a malfunctioning timer interrupt in QEMU if no new compare value was written within the ISR in tickless mode.

ibirnbaum commented on include/arch/arm/aarch32/cortex_a_r/timer.h:
```diff
@@ -63,6 +63,30 @@ static ALWAYS_INLINE void arm_arch_timer_set_compare(uint64_t val)
+
+static ALWAYS_INLINE uint8_t arm_arch_timer_get_int_status(void)
+{
+	return (uint8_t)(sys_read32(TIMER_REG_GET(TIMER_ISR)) & 0x1);
```
done, it's now called TIMER_ISR_EVENT_FLAG.

ibirnbaum commented on arch/arm/core/aarch32/prep_c.c:
```diff
@@ -182,7 +183,8 @@ void z_arm_prep_c(void)
 	z_bss_zero();
 	z_data_copy();
-#if defined(CONFIG_ARMV7_R) && defined(CONFIG_INIT_STACKS)
+#if (defined(CONFIG_ARMV7_R) && defined(CONFIG_INIT_STACKS)) \
```
done

ibirnbaum commented on arch/arm/core/aarch32/mmu/arm_mmu_priv.h:
```diff
@@ -0,0 +1,213 @@
+ * MMU initialization.
+ */
+struct arm_mmu_flat_range {
+	char *name;
```
done

evgeniy-paltsev commented:
@ibirnbaum may I ask you to rebase your changes on latest Zephyr version?
As I can see now from 'Files changed' tab you are deleting/changing a lot of ARC code which looks really strange :)

ibirnbaum commented:
@evgeniy-paltsev Something must have gone wrong yesterday. I had some conflicts due to certain files having been updated in the meantime, I merged, added & pushed the updated versions, the web frontend was happy with that, but the CI kept complaining that there was a merge conflict it couldn't resolve. I'm just trying to get rid of the merge commits that caused this issue. Not too sure about the chances of success, though...

ibirnbaum committed:
modules: cmsis: add core_ca inclusion for aarch32 Cortex-A SoCs

Add the inclusion of the Common Peripheral Access Layer for aarch32
Cortex-A SoCs in combination with the respective SoC's header file,
in which the feature flags evaluated in core_ca.h must be provided.

The corresponding CMSIS include path is added to the build by setting
the HAS_CMSIS_CORE_A Kconfig item for all aarch32 Cortex-A CPUs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@Weidmueller.com>

ibirnbaum committed:
tests: enable kernel test fatal/exception for aarch32 Cortex-A CPUs

Add consideration of aarch32 Cortex-A CPUs to a test case in which
architecture-specific assembly instructions are used in order to
explicitly trigger an exception. This test case already considers
aarch32 Cortex-R CPUs, the same instruction will be used by
Cortex-A CPUs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@Weidmueller.com>

ibirnbaum commented:
@jackrosenthal Thanks for cleaning up the exploding list of reviewers after that one broken merge that eventually got pushed onto my branch! To all those who didn't come for aarch32 stuff: move along, there's nothing to see here ;-)

The use of CMSIS in the MMU code is in progress, and initial runs of the kernel testsuite on the QEMU target didn't look too bad, but I did observe the occasional strange behaviour which I attribute to QEMU rather than the code, as it runs fine on actual hardware. It has been stated that the Xilinx targets integrated in vanilla QEMU aren't that great in terms of quality & stability, and that the way to go for such targets is the Xilinx derivative of QEMU in conjunction with a device tree rather than a pre-defined target. This is exactly how the Cortex-R5 QEMU board is set up, and therefore, the Xilinx derivative of QEMU is already a part of the SDK. Although branded arm64, this version also supports the Zynq, so I'm currently working on generating a device tree in order to have a reliable QEMU represenatation of the Cortex-A9.

ibirnbaum committed:
testsuite: enable timestamp_serialize() for aarch32 Cortex-A CPUs

Enable the existing aarch32 Cortex-R implementation of
timestamp_serialize() for aarch32 Cortex-A CPUs as well.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@Weidmueller.com>

ibirnbaum commented on arch/arm/core/aarch32/mmu/arm_mmu_priv.h:
```diff
@@ -0,0 +1,213 @@
+	uint32_t reserved	: 15;
+};
+
+static void arm_mmu_release_l2_table(struct arm_mmu_l2_page_table* l2_page_table);
```
done.

bbolen reviewed (changes requested):
Most of the stuff is nitpicking and I'm not necessarily asking for changes on those.  The mmu free table handling is the thing that I think needs to be looked at.

ibirnbaum committed:
cmake: boards: make QEMU icount sleep parameter configurable

When QEMU is running in icount mode, the behaviour regarding
the CPU being in sleep mode is configurable, with the default
value being set to 'off'. In this case, instruction counting
continues even when the CPU is sleeping or no target time is
set in the system timer. If the sleep parameter is set to
'on', instruction counting stops if the CPU is sleeping, and
the instruction count will fast-forward directly to the next
target time once it is being set.

In the Xilinx derivative of QEMU (which simulates ZynqMP and
Zynq-7000 SoCs), the default configuration causes QEMU to no
longer accept control commands via the console or remotely
whenever Zephyr enters and remains in idle mode.

Therefore, this commit makes the value of the sleep para-
meter of the icount functionality configurable, with the
default value remaining 'off'. The value can be changed at
board level alongside with the specification of the icount
shift value.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@Weidmueller.com>

carlescufi reviewed (approved):
This looks good to me. The only question I had revolved around having some sort of common `CORTEX_R_A32` Kconfig option to avoid the duplication of the `||`ing but honestly I cannot come up with a good name or description for it.

cfriedt commented:
@ibirnbaum, @carlescufi  - this might have introduced a new bug and regression into main. Is there a fix on the way?
```
twister -i -p qemu_cortex_a53_smp -s tests/kernel/fatal/exception/kernel.common.stack_sentinel
twister -i -p qemu_cortex_a53_smp  -s tests/kernel/smp/kernel.multiprocessing.smp
twister -i -p qemu_cortex_a53_smp  -s tests/kernel/sched/schedule_api/kernel.scheduler
twister -i -p qemu_cortex_a9 -s tests/kernel/fatal/exception/kernel.common.stack_sentinel
```
